File: test/ModularPipelines.UnitTests/NestedCollisionTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Exceptions;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;
using TUnit.Assertions.AssertConditions.Throws;

namespace ModularPipelines.UnitTests;

public class NestedCollisionTests
{
    [Test]
    public async Task Modules_Dependent_On_Each_Other_Throws_Exception()
    {
        await Assert.That(() => TestPipelineHostBuilder.Create()
                .AddModule<DependencyConflictModule1>()
                .AddModule<DependencyConflictModule2>()
                .AddModule<DependencyConflictModule3>()
                .AddModule<DependencyConflictModule4>()
                .AddModule<DependencyConflictModule5>()
                .ExecutePipelineAsync()).
            Throws<DependencyCollisionException>()
                .And.HasMessageEqualTo("Dependency collision detected: **DependencyConflictModule2** -> DependencyConflictModule3 -> DependencyConflictModule4 -> DependencyConflictModule5 -> **DependencyConflictModule2**");
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule2>]
    private class DependencyConflictModule1 : Module
    {
        protected override async Task<IDictionary<string, object>?> ExecuteAsync(IPipelineContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return null;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule3>]
    private class DependencyConflictModule2 : Module
    {
        protected override async Task<IDictionary<string, object>?> ExecuteAsync(IPipelineContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return null;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule4>]
    private class DependencyConflictModule3 : Module
    {
        protected override async Task<IDictionary<string, object>?> ExecuteAsync(IPipelineContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return null;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule5>]
    private class DependencyConflictModule4 : Module
    {
        protected override async Task<IDictionary<string, object>?> ExecuteAsync(IPipelineContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return null;
        }
    }

    [ModularPipelines.Attributes.DependsOn<DependencyConflictModule2>]
    private class DependencyConflictModule5 : Module
    {
        protected override async Task<IDictionary<string, object>?> ExecuteAsync(IPipelineContext context, CancellationToken cancellationToken)
        {
            await Task.CompletedTask;
            return null;
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/NestedCollisionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Exceptions
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module NestedCollisionTestModules =
    [<DependsOn(typeof<DependencyConflictModule2>)>]
    type DependencyConflictModule1() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule3>)>] DependencyConflictModule2() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule4>)>] DependencyConflictModule3() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule5>)>] DependencyConflictModule4() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

    and [<DependsOn(typeof<DependencyConflictModule2>)>] DependencyConflictModule5() =
        inherit Module<bool>()
        override _.ExecuteAsync(_, _) = Task.FromResult(true)

type NestedCollisionTests() =
    [<Test>]
    member _.Modules_Dependent_On_Each_Other_Throws_Exception() = async {
        let mutable message = null

        try
            let! _ =
                TestPipelineHostBuilder.Create()
                    .AddModule<NestedCollisionTestModules.DependencyConflictModule1>()
                    .AddModule<NestedCollisionTestModules.DependencyConflictModule2>()
                    .AddModule<NestedCollisionTestModules.DependencyConflictModule3>()
                    .AddModule<NestedCollisionTestModules.DependencyConflictModule4>()
                    .AddModule<NestedCollisionTestModules.DependencyConflictModule5>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
            ()
        with :? DependencyCollisionException as ex ->
            message <- ex.Message
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(message), "Dependency collision detected: **DependencyConflictModule2** -> DependencyConflictModule3 -> DependencyConflictModule4 -> DependencyConflictModule5 -> **DependencyConflictModule2**"))
    }
